flow/src/test/java/com/sequenceiq/flow/core/chain/FlowChainHandlerTest.java

```diff
@@ -0,0 +1,152 @@
+import static org.mockito.Mockito.when;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.flow.api.model.operation.OperationType;
+import com.sequenceiq.flow.core.FlowLogService;
+import com.sequenceiq.flow.core.cache.FlowStatCache;
+import com.sequenceiq.flow.core.chain.config.FlowChainOperationTypeConfig;
+import com.sequenceiq.flow.core.chain.config.FlowTriggerEventQueue;
+import com.sequenceiq.flow.domain.FlowChainLog;
+
+@ExtendWith(MockitoExtension.class)
+class FlowChainHandlerTest {
+
+    private static final String FLOW_CHAIN_ID = "flowchainId";
+
+    private static final String FLOW_CHAIN_TYPE = "type";
+
+    private static final String PARENT_FLOW_CHAIN_ID = "parentId";
+
+    private static final String USER_CRN = "userCrn";
+
+    private static final String TRIGGER_VALUE = "triggerValue";
+
+    private static final String SELECTOR = "simpleSelector";
+
+    private static final String SIMPLE_VALUE = "chainKey";
+
+    @Mock
+    private FlowChains flowChains;
+
+    @Mock
+    private FlowLogService flowLogService;
+
+    @Mock
+    private FlowStatCache flowStatCache;
+
+    @Mock
+    private FlowChainOperationTypeConfig flowChainOperationTypeConfig;
+
+    @InjectMocks
+    private FlowChainHandler underTest;
+
+    @Captor
+    private ArgumentCaptor<FlowTriggerEventQueue> eventQueueCaptor;
+
+    @Test
+    void testRestoreFlowChainNotFound() {
+        when(flowLogService.findFirstByFlowChainIdOrderByCreatedDesc(FLOW_CHAIN_ID)).thenReturn(Optional.empty());
+        underTest.restoreFlowChain(FLOW_CHAIN_ID);
+        verifyNoInteractions(flowChains, flowStatCache, flowChainOperationTypeConfig);
+    }
+
+    @ParameterizedTest(name = "use Jackson = {0}")
+    @ValueSource(booleans = { false, true })
+    void testRestoreFlowChain(boolean useJackson) {
+        FlowChainLog flowChainLog = new FlowChainLog(FLOW_CHAIN_TYPE, FLOW_CHAIN_ID, PARENT_FLOW_CHAIN_ID,
+                "{\"@type\":\"java.util.concurrent.ConcurrentLinkedQueue\",\"@items\":[{\"@type\":\"" +
+                        Simple.class.getName() + "\",\"resourceId\":123,\"key\":\"" + SIMPLE_VALUE + "\"}]}",
+                useJackson
+                        ? "[{\"@type\":\"" +
+                            Simple.class.getName() + "\",\"resourceId\":123,\"key\":\"" + SIMPLE_VALUE + "\"}]"
+                        : null,
```

Comment:
Sanity check: How can these be deserialized? The current impl of `Simple` does not have a field for `resourceId`, and its sole constructor expects a `selector` (besides the `key`) but not a `resourceId`.

Reply:
It is simple (pun intended).
ResourceId is omitted because there is no JsonProperty for it. Selector is not omitted but ignored in the constructor.
For clearness, I have refactored it a bit.